flatpak: init setup

siddarthkay opened the pull request:
fixes: https://github.com/status-im/status-app/issues/16341

## Notes

- [x] Depends on https://github.com/status-im/status-go/pull/7275

- `app.status.desktop.yml`: the flatpak manifest. Defines sandbox permissions, builds libcanberra/pcsc-lite/libusb/ccid as in-sandbox modules, and grants the `status-desktop` module read-only host access via `--filesystem=host:ro` and `--filesystem=host-os:ro` so its build script can read Qt, our compiled native libs, and the few host system libs we need without any pre-staging.

- `app.status.desktop.appdata.xml` : This is metadata for when we will push to a flatpak repository ( flathub ).

- `flatpak/0001-gtk-Don-t-assume-all-GdkDisplays-are-GdkX11Displays-.patch` : patch for libcanberra to work on non-X11 displays (Wayland).

- `status-512.png` - `flatpak` needs icon in this resolution, hence adding a new one seemed like the easiest solution. I initially tried generating a scaled up version of existing icon in repo but `flatpak` rejected it.

## How the flatpak is generated

`make flatpak` is a four-stage pipeline. Everything lands under `tmp/linux/flatpak/` and `pkg/`. The build sandbox reads what it needs straight from the host.

### Stage 0: prerequisites

These are built by the usual targets before `make flatpak` even starts. Same files the AppImage path consumes:

| Path | Built by |
|---|---|
| `bin/nim_status_client` | `make nim_status_client` (~88 MB) |
| `bin/StatusQ/libStatusQ.so` | `make statusq` (transitive) |
| `vendor/status-go/build/bin/libstatus.so.0` | `make status-go` (~124 MB) |
| `vendor/status-keycard-go/build/libkeycard/libkeycard.so` | `make status-keycard-go` |
| `${NIM_SDS_SOURCE_DIR}/build/libsds.so` | side effect of `make status-go` |
| `resources.rcc` | `make rcc` (transitive) |

### Stage 1: `scripts/bundle-flatpak.sh` runs `flatpak-builder` (~3 min)

Reads `FLATPAK_BUILD_DIR`, `FLATPAK_REPO_DIR`. Calls:

```
flatpak-builder --force-clean --disable-rofiles-fuse --disable-cache \
  --repo=$FLATPAK_REPO_DIR  $FLATPAK_BUILD_DIR  app.status.desktop.yml
```

flatpak-builder reads the manifest and processes each module in order, inside a sandbox built on `org.freedesktop.Sdk//24.08`:

| Module | Source | What lands in `/app` |
|---|---|---|
| `libcanberra` | upstream tarball + project-local patch | `lib/libcanberra*.so*`, `bin/canberra-gtk-play` |
| `pcsc-lite` | upstream tarball, meson build | `sbin/pcscd`, `lib/libpcsclite*.so*` (with `ipcdir=/tmp/pcscd/run` matching the wrapper) |
| `libusb` | upstream tarball | `lib/libusb-1.0.so*` |
| `ccid` | upstream tarball, meson build | `lib/pcsc/drivers/ifd-ccid.bundle` |
| `status-desktop` | enumerated `sources` + `--filesystem=host:ro` + `--filesystem=host-os:ro` | runs `./scripts/flatpak-copy-to-app.sh` |

The `status-desktop` module's `sources` list is the contract for what reaches the sandbox: only the project's own files (`bin/nim_status_client`, `bin/StatusQ/libStatusQ.so`, `resources.rcc`, `flatpak-wrapper.sh`, `status-512.png`, `scripts/flatpak-copy-to-app.sh`). The module's `build-options.build-args` adds two flatpak filesystem grants:

- `--filesystem=host:ro` mounts host's `/opt`, `/home`, `/mnt`, `/media`, `/run/media` at the same paths in the sandbox. That covers `/opt/qt` (Qt) and `/workspace` (our native libs in CI; also covers `$WORKSPACE_TMP/nim-sds`).
- `--filesystem=host-os:ro` mounts the host's `/usr`, `/bin`, `/sbin`, `/lib*` under `/run/host/` in the sandbox (it does not clobber the SDK's own `/usr`). That covers the krb5 chain and NSS PKCS#11 modules the runtime doesn't ship.

Inside the sandbox `scripts/flatpak-copy-to-app.sh` lays everything out:

```
/app/
├── bin/
│   ├── nim_status_client            the app
│   └── nim_status_client_wrapped    flatpak-wrapper.sh, the runtime entry point
├── sbin/pcscd
├── lib/
│   ├── libQt6*.so.6, libicu*, libav*           from /opt/qt (host)
│   ├── libstatus.so*, libkeycard.so, libsds.so from /workspace + $NIM_SDS (host)
│   ├── libkrb5*, libgssapi_krb5*, nss/         from /run/host/usr/lib (host-os)
│   ├── libcanberra*, libpcsclite*, libusb*, pcsc/drivers/  from manifest modules
│   ├── qt6/plugins/, qt6/qml/                  from /opt/qt (host)
│   └── debug/                                  split out into the .Debug ref
├── libexec/QtWebEngineProcess
├── resources/, resources.rcc, translations/
└── share/applications/, share/icons/
```

What's deliberately NOT shipped because the freedesktop runtime supplies it: GStreamer, libssl, libnss3 base, libcurl, libstdc++, X11 libs etc.

After every module is built, flatpak-builder commits the result as an OSTree commit and exports two refs into `$FLATPAK_REPO_DIR`:

```
tmp/linux/flatpak/repo/refs/heads/app/
├── app.status.desktop/x86_64/master
└── app.status.desktop.Debug/x86_64/master   (lib/debug/, exported as a separate ref)
```

### Stage 2: `flatpak build-bundle` (~30s)

Same script, second half. Pulls the `app.status.desktop` ref out of the OSTree repo and writes it as a single self-contained file:

```
flatpak build-bundle  $FLATPAK_REPO_DIR  $STATUS_CLIENT_FLATPAK  app.status.desktop
```

That gives us **`pkg/status-desktop.flatpak`** (~180 MB). A `.flatpak` is just an OSTree static-delta superblock, so `flatpak install` on the user's machine imports the objects into their local repo and creates a deployment.

If `LINUX_GPG_PRIVATE_KEY_FILE` is set the bundle gets signed by `scripts/sign-linux-file.sh`.

### Stage 3: runtime via `flatpak-wrapper.sh`

The manifest declares `command: nim_status_client_wrapped`, so flatpak's launcher executes the wrapper inside the sandbox. It:

1. Sets `LD_LIBRARY_PATH=/app/lib:/app/lib64`, `QT_PLUGIN_PATH=/app/lib/qt6/plugins`, `QML*_IMPORT_PATH=/app/lib/qt6/qml`, the `QTWEBENGINE*` paths, `GST_PLUGIN_PATH`, `GTK_PATH=/app/lib/gtk-3.0` (so GTK finds the bundled canberra module), `GTK_MODULES=canberra-gtk-module:...`.
2. If a previous run left a stale `pcscd` PID file, kills it and polls `kill -0` until the PID is actually gone (no blind sleep).
3. Starts `/app/sbin/pcscd --foreground --auto-exit` in the background and waits for `/tmp/pcscd/run/pcscd.comm` to show up before continuing.
4. `exec /app/bin/nim_status_client --dataDir="${HOME}/.status-im" "$@"`.

`--dataDir` is mandatory. Without it the app writes into the per-instance sandbox home, which gets wiped between runs.

### Make targets

| Target | Action |
|---|---|
| `make flatpak` | Stage 1 + Stage 2 (Stage 0 pulled in transitively), produces `pkg/status-desktop.flatpak` |
| `make flatpak-install` | builds, then `flatpak install --user --reinstall pkg/status-desktop.flatpak` |
| `make flatpak-run` | install, then `flatpak run app.status.desktop` |
| `make flatpak-clean` | wipes `tmp/linux/flatpak/{build-dir,repo}` and `pkg/*.flatpak`. Leaves `.flatpak-builder/downloads/` (cached upstream tarballs) alone, so the next build doesn't re-download libcanberra/pcsc-lite/libusb/ccid. |

### AppImage vs Flatpak sizes

| | AppImage | Flatpak |
|---|---|---|
| Bundled libs | 431 (linuxdeployqt walks the full `ldd` closure) | ~414 (Qt + native + 4 in-sandbox modules + krb5/nss) |
| Final size | 329 MB (squashfs) | 180 MB (OSTree static delta) |
| Why the delta | linuxdeployqt re-bundles glibc-adjacent libs (`libssl`, `libxcb-*`, `libwayland-*`, `libstdc++`, ...) | `org.freedesktop.Platform//24.08` provides them at install time |

## Screenshot of flatpak build running on my linux PC
<img width="1097" height="646" alt="Screenshot 2026-01-08 at 12 57 50 AM" src="https://github.com/user-attachments/assets/a8e1ad8f-78b8-4253-9082-4b2a86fce6fc" />

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/android/arm64/package/PR-19340#33](https://ci.status.im/job/status-desktop/job/prs/job/android/job/arm64/job/package/job/PR-19340/33/) :small_blue_diamond: ~12 min :small_blue_diamond: dc684278 :small_blue_diamond: [:package: android/arm64 package](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-UNKNOWN-251210-211551-dc6842-pr19340-arm64.apk)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | dc684278 | [#33](https://ci.status.im/job/status-desktop/job/prs/job/android/job/arm64/job/package/job/PR-19340/33/) | 2025-12-10 21:28:31 | ~12 min | `android/arm64` | [:robot:`apk`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-UNKNOWN-251210-211551-dc6842-pr19340-arm64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-UNKNOWN-251210-211551-dc6842-pr19340-arm64.apk&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :x: | 86ebeaf0 | [#34](https://ci.status.im/job/status-app/job/prs/job/android/job/arm64/job/package/job/PR-19340/34/) | 2025-12-12 05:28:49 | ~10 min | `android/arm64` | [:page_facing_up:`log`](https://ci.status.im/job/status-app/job/prs/job/android/job/arm64/job/package/job/PR-19340/34/consoleText) |

yakimant commented:
Related issues:
- https://github.com/status-im/status-app/issues/16341
- https://github.com/status-im/status-app/issues/8266

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -10,11 +10,11 @@ pipeline {
       args '--entrypoint="" ' +
-           '--cap-add=SYS_ADMIN ' +
-           '--security-opt=apparmor:unconfined ' +
+           '--privileged ' +
```
We should try to avoid this if possible.

jakubgs commented on app.status.desktop.yml:
```diff
@@ -0,0 +1,83 @@
+      - install -Dm644 nim-status.desktop /app/share/applications/app.status.desktop.desktop
+      - sed -i 's|^Exec=.*|Exec=nim_status_client_wrapped|;s|^Icon=.*|Icon=app.status.desktop|' /app/share/applications/app.status.desktop.desktop
+      - install -Dm644 status-512.png /app/share/icons/hicolor/512x512/apps/app.status.desktop.png
+      - install -Dm644 app.status.desktop.appdata.xml /app/share/metainfo/app.status.desktop.appdata.xml
```
Aren't we just kinda repeating what we do in `scripts/init_app_dir.sh` here but slightly differently?

jakubgs commented on app.status.desktop.appdata.xml:
```diff
@@ -0,0 +1,43 @@
+    </ul>
+  </description>
+  <launchable type="desktop-id">app.status.desktop.desktop</launchable>
+  <url type="homepage">https://status.im</url>
```
Do not reference old `status.im` domain.

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -168,10 +200,7 @@ pipeline {
     failure { script { github.notifyPR(false) } }
-    cleanup {
-      cleanWs(disableDeferredWipeout: true)
-      dir(env.WORKSPACE_TMP) { deleteDir() }
```
Why are we removing this?

siddarthkay replied:
my bad was removed by mistake.

jakubgs commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,39 @@
+done
+
+# Launch application
+exec /app/bin/nim_status_client --dataDir="${HOME}/.status-im" "$@"
```
Why is the `dataDir` setting necessary? Shouldn't that be the default?

siddarthkay replied:
The default value in case of `flatpak` lands in some etc space which crashed the app at startup. Hence I needed to add this, I'll add a comment to explain

jakubgs replied:
Okay, that's dumb, but may be worth a comment.

jakubgs commented on scripts/build-flatpak.sh:
```diff
@@ -0,0 +1,86 @@
+      sys-libs/ 2>/dev/null || true
+cp -P /usr/local/lib/x86_64-linux-gnu/libpcsclite*.so* sys-libs/ 2>/dev/null || true
+cp -r /usr/lib/x86_64-linux-gnu/nss sys-libs/ 2>/dev/null || true
+cp -r /usr/lib/x86_64-linux-gnu/gstreamer-1.0 sys-libs/ 2>/dev/null || true
```
This script is also kinda like a slightly different version of `scripts/init_app_dir.sh`. Are we sure we want to maintain those separately?

siddarthkay replied:
I'll check again.

jakubgs reviewed (commented):
I like how you organized things in `scripts/lib-deps.sh`.

yakimant reviewed (changes requested):
2 main issues with structure imho:
- init_app_dir.sh should be called for flatpak too?
- functions from lib-deps.sh should copy to DEST without staging dirs

siddarthkay commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,69 @@
+    "$pcscd" --foreground --auto-exit &
+    PCSCD_PID=$!
+    echo "$PCSCD_PID" > "$PCSCD_PIDFILE"
+    sleep 0.3
```
added a comment, the number is random wait for smartcard daemon to start.
I can experiment with this number further.

siddarthkay commented on Makefile:
```diff
@@ -806,6 +806,32 @@ ifdef LINUX_GPG_PRIVATE_KEY_FILE
+export FLATPAK_REPO_DIR ?= tmp/flatpak-repo
+
+# Build flatpak using the CI script
+.PHONY: flatpak
```
fixed.

yakimant reviewed (changes requested):
I finally went deep into the code. Sorry didnt do it before.

I don't like the way code is structured, it's a bit of spagetti.

My suggestion:

**Dont branch in init_app_dir.sh, leave it for appimage bundling**

AppDir is a format for AppImage preparation dir:
https://docs.appimage.org/reference/appdir.html

2 branches you have a large, not shared and confusing.
For example, why `flatpak` branch ueses libs for qt libs, but not `appimage`?
Why `appimage` uses qtwebengine copy from libs, but not `flatpack`

**Remove non-shared functions from lib-deps.sh**

`lib-deps.sh` is generally a good idea.

But, some functions are shared between `appimage` and `flatpak`, but some not.
For example, why `flatpak` copies qt files using `lib-deps`, but appimage not?
Why `appimage` copies qtwebengine with `lib-deps`, but not `flatpak`?

**Clear flow for `flatpak` bundling**

1. `stage-flatpak-deps.sh`: prepare/stage libs for flatpak

This is a part of `build-flatpak.sh`, which copies the libs.

2. `bundle-flatpak.sh: bundles `flatpak` from files

This is a part of `build-flatpak.sh` running `flatpak` tools:
- `flatpak-builder` to prepare the `repo`
- `flatpak build-bundle` to export a `flatpak` from `repo`

3.  `flatpak-copy-to-app.sh`: copies files to /app

This is what you have in `init_app_dir.sh` `flatpak` branch now.
As I said - leave `init_app_dir.sh` for `AppDir`.

**Notes**

- I dont understand, what inside `build-dir` and `repo` folders. I also dont understand why do you need staging, cant the files be just copied by manifest script? If not - what the purpose of `build-dir`, maybe its a good place for staging?
- In suggested structure you will have a clear picture of what each script is doing, making a flow of different sel-sufficient steps with inputs/outputs (build-dir, repo, staged libs, flatpak); not mess with appdir script, except of shared code in lib-deps.sh
- Names of sh files I proposed are not great, maybe you have a better name
- You can join scripts 1 and 2 into one script, but then make functions for each part of functionality
- `init_flatpak_build_dir.sh` and `init_app_dir.sh can share more code if needed through `lib-deps.sh`

yakimant commented on scripts/build-flatpak.sh:
```diff
@@ -0,0 +1,102 @@
+# Configuration (can be overridden via environment)
+FLATPAK_MANIFEST="${FLATPAK_MANIFEST:-app.status.desktop.yml}"
+FLATPAK_BUILD_DIR="${FLATPAK_BUILD_DIR:-tmp/flatpak-build}"
+FLATPAK_REPO_DIR="${FLATPAK_REPO_DIR:-tmp/flatpak-repo}"
```
Also lets define only in Makefile

yakimant commented:
@siddarthkay things to consider
- [Hosting a repository](https://docs.flatpak.org/en/latest/hosting-a-repository.html) later if Flathub will be too difficult
- Can we share system libs to build sandbox environement instead of staging workaround? https://docs.flatpak.org/en/latest/sandbox-permissions.html

micieslak reviewed (commented):
Nice work! I manged to build the package, install and run it:

<img width="1801" height="930" alt="Image" src="https://github.com/user-attachments/assets/5640a37e-5365-489b-9860-1a4ff8ace530" />

However I encountered some problems I had to resolve first:

- installing flatpak and flatpak runtimes. It would be nice to have a list of prerequisites that needs to be done on fresh system to successfully run `make flatpak`
-  there is Qt path hardcoded. I had to modify `bundle-flatpak.sh` and `flatpak-copy-to-app.sh` to run the command with specific qt dir: `QT_DIR=/home/michalc/Qt/6.11.0/gcc_64 make flatpak`. It should be possible to indicate custom Qt location out of the box imo.

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -124,28 +129,58 @@ pipeline {
+          steps { script {
+            status.buildSigned(platform='linux', target='flatpak')
+          } }
+        }
```
Why shouldn't the Flatpak build not be a separate job?

yakimant replied:
We need to build once, pack 2 times (appimage, flatpak).
Previously it was 2 copy-paste jobs with build duplication.

I suggested Sid to reuse build results instead of building the same thing twice.

There are multiple ways to achive that:
1. 3 jobs (build and 2 dependent pacaging jobs)
2. 1 job with packing for appimage and flatpak in parallel (like Sid did)

siddarthkay replied:
moved back to 2 jobs, linux e2e was failing due to some issue with build type not being release.

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -161,19 +196,34 @@ pipeline {
   }
 }
 
+def notifyPRBoth(success) {
```
Unnecessary complication, lets just run separate jobs.

yakimant commented on ci/Jenkinsfile.linux:
```diff
@@ -124,28 +129,58 @@ pipeline {
       }
     }
 
+    stage('nim_status_client') {
```
```suggestion
    stage('Build') {
```

yakimant commented on app.status.desktop.yml:
```diff
@@ -0,0 +1,158 @@
+    buildsystem: simple
+    build-options:
+      build-args:
+        # `host` covers /opt and the user's home (workspace lives there) so
+        # the script can read /opt/qt and /workspace directly.
+        - --filesystem=host:ro
+        # `host-os` exposes the host's /usr, /bin, /sbin, /lib* read-only at
+        # /run/host inside the sandbox. The freedesktop runtime doesn't ship
+        # the krb5 chain or NSS's PKCS#11 modules, so we read them from there.
+        - --filesystem=host-os:ro
+        # /tmp itself is bwrap-managed (mounting it ro breaks the sandbox), so
+        # we narrow-mount only the nim-sds dir for local-dev. In CI this is a
+        # no-op: NIM_SDS_SOURCE_DIR sits under WORKSPACE_TMP under home.
+        - --filesystem=/tmp/nim-sds:ro
```
Amazing! Happy it worked!

yakimant commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,90 @@
+# a fixed delay. $1=description $2=timeout-seconds $3...=command.
+wait_for() {
+  local desc="$1" timeout="$2"; shift 2
+  local deadline=$(( SECONDS + timeout ))
```
Nice! Didnt know about `SECONDS`

yakimant commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,90 @@
+      echo "WARN: timed out after ${timeout}s waiting for ${desc}" >&2
+      return 1
+    fi
+    sleep 0.05
```
I think 0.05 is too small, how about timeout/20?

yakimant commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,90 @@
+  local deadline=$(( SECONDS + timeout ))
+  while ! "$@"; do
+    if (( SECONDS >= deadline )); then
+      echo "WARN: timed out after ${timeout}s waiting for ${desc}" >&2
```
Timeout is killing the app? Should be ERROR or FATAL then.

yakimant commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,90 @@
+rm -rf "$PCSCD_RUN_DIR"
+mkdir -p "$PCSCD_RUN_DIR" "$(dirname "$PCSCD_PIDFILE")" "$PCSCLITE_CONFIG_DIR"
+
+for pcscd in /app/sbin/pcscd /app/bin/pcscd; do
```
Really 2 needed?

yakimant commented on flatpak-wrapper.sh:
```diff
@@ -0,0 +1,90 @@
+      bash -c "! kill -0 $OLD_PID 2>/dev/null"
+  fi
+  rm -f "$PCSCD_PIDFILE"
+fi
```
Is it just a leftover file and no proccess should be running?
Did you see situations proccess continued to run after a crash?

yakimant commented on scripts/flatpak-copy-to-app.sh:
```diff
@@ -0,0 +1,70 @@
+
+# Qt path. Keep the version below in sync with ci/Dockerfile's ARG QT_VERSION
+# and ci/Jenkinsfile.
+QT_SRC="/opt/qt/6.11.0/gcc_64"
```
No way you could pass it from outside?

yakimant commented on scripts/flatpak-copy-to-app.sh:
```diff
@@ -0,0 +1,70 @@
+Icon=app.status.desktop
+Type=Application
+Categories=Network;InstantMessaging;Chat;
+DESKTOP
```
I would save it as a repo file explicitly.

yakimant commented on scripts/init_app_dir.sh:
I would make a separate commit if not related to flatpak.

yakimant reviewed (approved):
Mostly fine by me, we can polish it forever
- Left some minor comments
- If @jakubgs is happy with Jenkinsfile or there is some change
- Would be nice to keep unrelated changes separately (like appimage init_app_dir) - it makes reviewes harder and also you need to test that appimage is not broken

siddarthkay committed:
refactor(linux): tidy AppDir assembly in init_app_dir.sh

Extract copy_native_libs/copy_system_libs/copy_qt_webengine helpers and
drive the Qt path from $QTDIR.

siddarthkay committed:
flatpak: init setup
